packages-ts/starknet-gauntlet-ocr2/src/lib/encoding.ts

```diff
@@ -20,46 +42,64 @@ export function bytesToFelts(data: Uint8Array): string[] {
   return felts
 }
 
+const MAX_LEN: bigint = (BigInt(1) << BigInt(54)) - BigInt(1)
+
 export function feltsToBytes(felts: string[]): Buffer {
-  console.log(felts)
-  let data = []
+  const data: number[] = []
 
-  // TODO: validate len > 1
+  if (!felts.length) {
+    throw new Error('Felt string is empty')
+  }
 
-  // TODO: validate it fits into 54 bits
-  let length = Number(BigInt(felts.shift()))
+  const remainingLengthBigInt = BigInt(felts.shift())
+  if (remainingLengthBigInt > MAX_LEN) {
+    throw new Error('Length does not fit in 54 bits')
+  }
 
-  for (const felt of felts) {
-    let size = Math.min(CHUNK_SIZE, length)
-    let chunk = Uint8Array.from(Buffer.from(felt, 'hex'))
-    let filler = chunk.length - size
-    // pad to size
-    while (filler > 0) {
-      data.push(0)
+  let remainingLength = Number(remainingLengthBigInt)
+  for (let felt of felts) {
+    if (remainingLength <= 0) {
+      throw new Error(
+        `Too many felts (${felts.length}) for length ${remainingLengthBigInt.toString()}`,
+      )
     }
-    data.push(...chunk)
-
-    length -= chunk.length
+    const chunkSize = Math.min(CHUNK_SIZE, remainingLength)
+    let packedValue: bigint = BigInt(felt)
+    const unpackedValues: number[] = []
+    for (let i = 0; i < chunkSize; i++) {
+      unpackedValues.push(Number(packedValue & BigInt(0xff)))
+      packedValue = packedValue >> BigInt(8)
+    }
+    unpackedValues.reverse()
+    data.push(...unpackedValues)
+    remainingLength -= chunkSize
   }
 
   return Buffer.from(data)
 }
 
 export const decodeOffchainConfigFromEventData = (data: string[]): encoding.OffchainConfig => {
-  const oraclesLen = Number(num.toBigInt(data[3]))
-  /** SetConfig event data has the following info:
```

Comment:
These docs still seem useful to peek into the structure. Consider adding similar docs to the new version.